python/ray/data/impl/table_block.py

```diff
@@ -0,0 +1,231 @@
+import collections
+
+from typing import Dict, Iterator, List, Union, Tuple, Any, TypeVar
+
+from ray.data.block import Block, BlockAccessor
+from ray.data.impl.block_builder import BlockBuilder
+from ray.data.impl.simple_block import SimpleBlockBuilder
+from ray.data.impl.size_estimator import SizeEstimator
+
+T = TypeVar("T")
+
+# An Arrow block can be sorted by a list of (column, asc/desc) pairs,
+# e.g. [("column1", "ascending"), ("column2", "descending")]
+SortKeyT = List[Tuple[str, str]]
+GroupKeyT = Union[None, str]
+
+# The max size of Python tuples to buffer before compacting them into a
+# table in the BlockBuilder.
+MAX_UNCOMPACTED_SIZE_BYTES = 50 * 1024 * 1024
+
+
+class TableRow:
+    def __init__(self, row: Any):
+        self._row = row
+
+    def as_pydict(self) -> dict:
+        raise NotImplementedError
+
+    def keys(self) -> Iterator[str]:
+        return self.as_pydict().keys()
+
+    def values(self) -> Iterator[Any]:
+        return self.as_pydict().values()
+
+    def items(self) -> Iterator[Tuple[str, Any]]:
+        return self.as_pydict().items()
+
+    def __getitem__(self, key: str) -> Any:
+        raise NotImplementedError
+
+    def __eq__(self, other: Any) -> bool:
+        return self.as_pydict() == other
+
+    def __str__(self):
+        return str(self.as_pydict())
+
+    def __repr__(self):
+        return str(self)
+
+    def __len__(self):
+        raise NotImplementedError
+
+
+class DelegatingBlockBuilder(BlockBuilder[T]):
+    def __init__(self):
+        self._builder = None
+
+    def add(self, item: Any) -> None:
+        from ray.data.impl.arrow_block import ArrowRow, ArrowBlockBuilder
+        from ray.data.impl.pandas_block import PandasRow, PandasBlockBuilder
```

Comment:
I tried, but since this is the base file, moving these imports to the beginning of this file will create recursive imports, which results in

```
ImportError while loading conftest '/Users/yangkai/dev/github/alipay/ray/python/ray/data/tests/conftest.py'.
conftest.py:6: in <module>
    import ray
../../__init__.py:115: in <module>
    from ray import data  # noqa: E402,F401
../__init__.py:1: in <module>
    from ray.data.read_api import from_items, range, range_arrow, \
../read_api.py:20: in <module>
    from ray.data.dataset import Dataset
../dataset.py:30: in <module>
    from ray.data.datasource import (
../datasource/__init__.py:1: in <module>
    from ray.data.datasource.datasource import (Datasource, RangeDatasource,
../datasource/datasource.py:11: in <module>
    from ray.data.impl.arrow_block import ArrowRow
../impl/arrow_block.py:14: in <module>
    from ray.data.impl.table_block import TableBlockAccessor, TableRow, \
../impl/table_block.py:9: in <module>
    from ray.data.impl.arrow_block import ArrowRow, ArrowBlockBuilder
E   ImportError: cannot import name 'ArrowRow' from 'ray.data.impl.arrow_block' (/Users/yangkai/dev/github/alipay/ray/python/ray/data/impl/arrow_block.py)
```

Reply:
Hmm we should figure out how to resolve this, maybe some code needs to be moved.

Imports in the inner loop is very bad for performance

Reply:
I've moved `DelegatingBlockBuilder` to a separate file. Problem solved.